### ISSUE TITLE
certificate_request_context SHALL be zero length outside of PHA not enforced by client

### ISSUE THREAD

**Issue by t184256** [openssl/openssl#18965]
[RFC 8446 section 4.3.2](https://datatracker.ietf.org/doc/html/rfc8446#section-4.3.2) imposes the following restrictions on certificate_request_context:

> An opaque string which identifies the certificate request and which will be echoed in the client's Certificate message.  The certificate_request_context MUST be unique within the scope of this connection (thus preventing replay of client CertificateVerify messages).  **This field SHALL be zero length unless used for the post-handshake authentication exchanges** described in [Section 4.6.2](https://datatracker.ietf.org/doc/html/rfc8446#section-4.6.2).

OpenSSL stores certificate_request_context
https://github.com/openssl/openssl/blob/fc5888ccb60f33b366972299f30b976c4dc12162/ssl/statem/statem_clnt.c#L2385
and echoes it back
https://github.com/openssl/openssl/blob/fc5888ccb60f33b366972299f30b976c4dc12162/ssl/statem/statem_clnt.c#L3550
without performing a client-side check for the aforementioned SHALL clause.

As a result, against a non-compliant server sending such a non-PHA non-empty certificate_request_context, OpenSSL behaviour is less strict than [GnuTLS](https://gitlab.com/gnutls/gnutls/-/blob/df6c95f7d156100bd935c6904d588b0aef6e3442/lib/tls13/certificate_request.c#L151) and [NSS](https://hg.mozilla.org/projects/nss/file/6d025c93055451074cf975089701faf3f221e519/lib/ssl/tls13con.c#l2752) one, which is to abort a connection.

**Comment by t8m** [comment]
IMO this is not a strict compliance issue as the RFC does not say the client MUST check this.

**Comment by t184256** [comment]
True, even though it's an "absolute requirement", it's not spelled out explicitly that the client must abort (and how), so it's a question of consistency across implementations in an underdefined non-compliant situation.

**Comment by tomato42** [comment]
first of all: [RFC 2119 section 1](https://www.rfc-editor.org/rfc/rfc2119#section-1):
```
 1. MUST   This word, or the terms "REQUIRED" or "SHALL", mean that the
   definition is an absolute requirement of the specification.
```
Second of all, the RFC explicitly states that the uniqueness of the context is a requirement for resistance against replay attacks. As such, not checking it may have security implications, even if they're not explicitly stated.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the end of 3.4 dev barring further input
